pkg/rbd/nodeserver.go

```diff
@@ -378,126 +411,73 @@ func (ns *NodeServer) NodeUnstageVolume(ctx context.Context, req *csi.NodeUnstag
-	}
-	// Unmount the volume
-	devicePath, cnt, err := mount.GetDeviceNameFromMount(ns.mounter, stagingTargetPath)
-	if err != nil {
-		return nil, status.Error(codes.Internal, err.Error())
 	}
 
-	if err = ns.unmount(stagingTargetPath, devicePath, cnt); err != nil {
-		return nil, err
-	}
-
-	if stagingTargetPath == blockStagingPath {
-		if err = os.Remove(stagingTargetPath); err != nil {
+	if err = os.Remove(stagingTargetPath); err != nil {
+		// Any error is critical as Staging path is expected to be empty by Kubernetes, it otherwise
+		// keeps invoking Unstage. Hence any errors removing files within this path is a critical
+		// error
+		if !os.IsNotExist(err) {
+			klog.Errorf("failed to remove staging target path (%s): (%v)", stagingTargetPath, err)
 			return nil, status.Error(codes.Internal, err.Error())
 		}
 	}
 
-	klog.Infof("rbd: successfully unmounted volume %s from %s", req.GetVolumeId(), stagingTargetPath)
-
-	return &csi.NodeUnstageVolumeResponse{}, nil
-}
-
-func (ns *NodeServer) unmount(targetPath, devicePath string, cnt int) error {
-	var err error
-	// Bind mounted device needs to be resolved by using resolveBindMountedBlockDevice
-	if devicePath == "devtmpfs" {
-		devicePath, err = resolveBindMountedBlockDevice(targetPath)
-		if err != nil {
-			return status.Error(codes.Internal, err.Error())
+	imgInfo, err := lookupRBDImageMetadataStash(stagingParentPath)
+	if err != nil {
+		klog.V(2).Infof("failed to find image metadata: %v", err)
+		// It is an error if it was mounted, as we should have found the image metadata file with
+		// no errors
+		if !notMnt {
+			return nil, status.Error(codes.Internal, err.Error())
 		}
-		klog.V(4).Infof("NodeUnpublishVolume: devicePath: %s, (original)cnt: %d\n", devicePath, cnt)
-		// cnt for GetDeviceNameFromMount is broken for bind mouted device,
-		// it counts total number of mounted "devtmpfs", instead of counting this device.
-		// So, forcibly setting cnt to 1 here.
-		// TODO : fix this properly
-		cnt = 1
-	}
 
-	klog.V(4).Infof("NodeUnpublishVolume: targetPath: %s, devicePath: %s\n", targetPath, devicePath)
+		// If not mounted, and error is anything other than metadata file missing, it is an error
+		if _, ok := err.(ErrMissingStash); !ok {
+			return nil, status.Error(codes.Internal, err.Error())
+		}
 
-	// Unmounting the image
-	err = ns.mounter.Unmount(targetPath)
-	if err != nil {
-		klog.V(3).Infof("failed to unmount targetPath: %s with error: %v", targetPath, err)
-		return status.Error(codes.Internal, err.Error())
+		// It was not mounted and image metadata is also missing, we are done as the last step in
+		// in the staging transaction is complete
+		return &csi.NodeUnstageVolumeResponse{}, nil
 	}
 
-	cnt--
-	if cnt != 0 {
-		// TODO should this be fixed not to success, so that driver can retry unmounting?
-		return nil
-	}
+	// Find device mapped to image
+	devicePath, found := findDeviceMappingImage(imgInfo.Pool, imgInfo.ImageName, imgInfo.NbdAccess)
```

Comment:
as we know the devicePath after rbd map we can add the devicePath to the json file?

Reply:
True, it can be done, BUT I did not want to break the transaction order, and hence add or assume we will find this later in the image metadata stash. IOW, if partway through the transaction errors or faults occur, we will not have that information to stash, and hence will need to fall back to other means.

Reply:
what transaction order?

what are the transaction errors can happen here?

Reply:
Read this comment: https://github.com/ceph/ceph-csi/pull/510/files#diff-a2c76a9049f52b0f018e9dac09c35f0cR49-R54

IOW, we create the `config.json` first before mapping the device, so we are sure to find the file to help with unmap and other operations as required. Hence, updating it (with the device name) post the fact, brings in an assumption in the code that this may or may not have happened, which I want to avoid.

Reply:
Another consideration here would be to use the `imagespec` as we know the `pool` and `imagename` from the stashed data, instead of determining the devicePath at all, as pointed out by @Madhu-1 [here](https://github.com/ceph/ceph-csi/issues/520#issuecomment-518109000)

I am looking at this (to ensure we get errors that help when the image is not mapped, to detect and ignore the same), to refrain from attempting to get the `devicePath` at all in the first place

Reply:
> I am looking at this (to ensure we get errors that help when the image is not mapped, to detect and ignore the same), to refrain from attempting to get the `devicePath` at all in the first place

The CLI output when the images are not mapped like as below,
```
# rbd unmap --device-type krbd replicapool/csi-vol-7101dd25-b7c3-11e9-8b61-d252735dc0f9
rbd: replicapool/csi-vol-7101dd25-b7c3-11e9-8b61-d252735dc0f9: not a mapped image or snapshot
rbd: unmap failed: (22) Invalid argument

# rbd unmap --device-type nbd replicapool/csi-vol-7101dd25-b7c3-11e9-8b61-d252735dc0f9
rbd-nbd: replicapool/csi-vol-7101dd25-b7c3-11e9-8b61-d252735dc0f9 is not mapped
rbd: rbd-nbd failed with error: /usr/bin/rbd-nbd: exit status: 1
```

We could use `imagespec` and look for these errors as benign and proceed, or leave the code as is for now, fixing the unmap problem first. @Madhu-1 thoughts?

Reply:
if you want devicePath during unmapping. why can't we just store the devicePath in metadata and use it for unmapping? 
the above error can be discarded as the image will not be mapped at all

Reply:
As answered [here](https://github.com/ceph/ceph-csi/pull/510#discussion_r310554247) and [here](https://github.com/ceph/ceph-csi/pull/510#discussion_r310560683), I do not want to touch the json file again post creating it.

Further we can also avoid determining which device file it is as noted [here](https://github.com/ceph/ceph-csi/pull/510#discussion_r310624209), if we use `imagespec` from the saved information. Which is an open option for now.

I want to follow a strict order, such that we do not update the file post creating it. In case of plugin restarts in between a Stage request, recovery cannot rely on device name in the file, as that may not have been updated post mapping the image. I hence do not want to add add the device path to the file, but would rather determine it as in the code before or use `imagespec`.

Reply:
Discussed this off-line with @Madhu-1 and decided to optimize out the call to get device lists and detect the `devicePath` to unmap, and instead use the `imagespec` version of `rbd unmap` command to achieve the same.

Reply:
Fixed as per last comment.